From 9b22ea17d02ed80fd7cceb44baeb13427184e9a3 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:10:06 -0400
Subject: [powerpc] mm: Make sure a local_irq_disable prevent a parallel THP split

Message-id: <1410545655-205645-378-git-send-email-dzickus@redhat.com>
Patchwork-id: 94498
O-Subject: [RHEL7 PATCH 377/626] powerpc/mm: Make sure a local_irq_disable prevent a parallel THP split
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 346519a1606e196bd5f11b9f5eacaf60f70477a1
Author: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
Date:   Sat Mar 15 16:17:58 2014 +0530

    powerpc/mm: Make sure a local_irq_disable prevent a parallel THP split

    We have generic code like the one in get_futex_key that assume that
    a local_irq_disable prevents a parallel THP split. Support that by
    adding a dummy smp call function after setting _PAGE_SPLITTING. Code
    paths like get_user_pages_fast still need to check for _PAGE_SPLITTING
    after disabling IRQ which indicate that a parallel THP splitting is
    ongoing. Now if they don't find _PAGE_SPLITTING set, then we can be
    sure that parallel split will now block in pmdp_splitting flush
    until we enables IRQ

    Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
    Acked-by: Rik van Riel <riel@redhat.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/mm/pgtable_64.c b/arch/powerpc/mm/pgtable_64.c
index 62bf5e8..f6ce1f1 100644
--- a/arch/powerpc/mm/pgtable_64.c
+++ b/arch/powerpc/mm/pgtable_64.c
@@ -647,6 +647,11 @@ void pmdp_splitting_flush(struct vm_area_struct *vma,
   if (old & _PAGE_HASHPTE)
    hpte_do_hugepage_flush(vma->vm_mm, address, pmdp);
  }
+ /*
+  * This ensures that generic code that rely on IRQ disabling
+  * to prevent a parallel THP split work as expected.
+  */
+ kick_all_cpus_sync();
 }
 
 /*
-- 
1.7.1